python/src/sdf/pyParam.cc

```diff
@@ -0,0 +1,176 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "pyParam.hh"
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <memory>
+#include <string>
+
+#include "pybind11_helpers.hh"
+#include "sdf/Element.hh"
+#include "sdf/Param.hh"
+#include "sdf/PrintConfig.hh"
+#include "sdf/config.hh"
+
+using namespace pybind11::literals;
+namespace py = pybind11;
+namespace sdf
+{
+// Inline bracket to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE {
+namespace python
+{
+
+void defineParam(py::object module)
+{
+  using PyClassParam = py::class_<Param, ParamPtr>;
+  auto paramClass = PyClassParam(module, "Param");
+
+  paramClass
+      .def(py::init(
+               [](const std::string &_key, const std::string &_typeName,
+                  const std::string &_default, bool _required,
+                  const std::string &_description)
+               {
+                 Errors errors;
+                 auto output = Param(_key, _typeName, _default, _required,
+                                          errors, _description);
+                 ThrowIfErrors(errors);
+                 return output;
+               }),
+           "key"_a, "type_name"_a, "default"_a, "required"_a,
+           "description"_a = "")
+      .def(py::init(
+               [](const std::string &_key, const std::string &_typeName,
+                  const std::string &_default, bool _required,
+                  const std::string &_minValue, const std::string &_maxValue,
+                  const std::string &_description)
+               {
+                 Errors errors;
+                 auto output =
+                     Param(_key, _typeName, _default, _required, _minValue,
+                                _maxValue, errors, _description);
+                 ThrowIfErrors(errors);
+                 return output;
+               }),
+           "key"_a, "type_name"_a, "default"_a, "required"_a, "min_value"_a,
+           "max_value"_a, "description"_a = "")
+      .def(py::init<const Param &>())
+      .def("get_as_string",
+           ErrorWrappedCast<const PrintConfig &>(&Param::GetAsString,
+                                                 py::const_),
+           "Get the value as a string.", "config"_a = PrintConfig())
+      .def("get_default_as_string",
+           ErrorWrappedCast<const PrintConfig &>(&Param::GetDefaultAsString,
+                                                 py::const_),
+           "Get the default value as a string", "config"_a = PrintConfig())
+      .def("get_min_value_as_string",
+           ErrorWrappedCast<const PrintConfig &>(&Param::GetMinValueAsString,
+                                                 py::const_),
+           "Get the minimum allowed value as a string.",
+           "config"_a = PrintConfig())
```

Comment:
There are couple of implementations for all of these methods, do we want to bind them?

Reply:
I think I've created bindings for all the ones that have `sdf::Errors`.